langserver/handler.go

```diff
@@ -21,9 +21,33 @@ import (
 
 // NewHandler creates a Go language server handler.
 func NewHandler() jsonrpc2.Handler {
-	return jsonrpc2.AsyncHandler(jsonrpc2.HandlerWithError((&LangHandler{
+	return lspHandler{jsonrpc2.HandlerWithError((&LangHandler{
 		HandlerShared: &HandlerShared{},
-	}).handle))
+	}).handle)}
+}
+
+// lspHandler wraps LangHandler to correctly handle requests in the correct
+// order.
+//
+// The LSP spec dictates a strict ordering that requests should only be
+// processed serially in the order they are received. However, implementations
+// are allowed to do concurrent computation if it doesn't affect the
+// result. We actually can return responses out of order, since vscode does
+// not seem to have issues with that. We also do everything concurrently,
+// except methods which could mutate the state used by our typecheckers (ie
+// textDocument/didOpen, etc). Those are done serially since applying them out
+// of order could result in a different textDocument.
+type lspHandler struct {
+	jsonrpc2.Handler
+}
+
+// Handle implements jsonrpc2.Handler
+func (h lspHandler) Handle(ctx context.Context, conn *jsonrpc2.Conn, req *jsonrpc2.Request) {
+	if isFileSystemRequest(req.Method) {
```

Comment:
It kind of acts like changes to the filesystem at the moment, it will see whatever the state of the FS at the time we do read. However, this guarantees exactly what we need to give the correct behaviour to the end user:

* incremental FS mutations are processed in the correct order.
* a reference request after a FS mutation will see the at least that version (same as if it was backed by a mutable FS)

Technically an older request that is still running may see state from a later FS mutation. However, I believe this is fine since that is what is in the users editor + vscode has likely already canceled the older request.

Making it so that non-mutable requests have a snapshot of the overlay FS is possible to give the full guarantee. However, I don't believe it is necessary + is more complicated + is impossible if it was backed by an FS anyways.